Add BDD page to QA section (#62)
UKHO/docs: docs/quality-assurance/bdd.md

> # Behavioural-Driven Design (BDD) at UKHO

Reviewer: BDD stands for Behaviour-driven development?
```suggestion
# Behaviour-Driven Development (BDD) at UKHO
```

Author: Quite right!

> "An approach that enhances communication between stakeholders and project team members by expressing product needs as concrete examples"

This page contains information about BDD and how it can be used to aid how we develop software at UKHO.

## What is BDD

* A process to enable the entire team to better understand stories before they are played
* BDD is about _collaboration_ - Product Owners (proxy for the customer), developers and testers communicating to gain an agreed understanding
* Uses rules and examples to illustrate this understanding
* When stories are played there is less ambiguity leading to better software development
* BDD is NOT:
  * Testing
  * SpecFlow or Cucumber
  * Just automated tests (tests are a downstream benefit)

## Aims of BDD

BDD aims to make the development of software more successful by:

* Incorporating diverse perspectives when defining requirements
* Creating a shared understanding in a delivery team of the required product changes
* Reducing the amount of re-work required in the maintenance phase (by making the original change more complete)
* Simplifying and shortening team Scrum events (refinement and planning)

> 

Reviewer: Might be worth adding something around helps ensure only code relevant to the behaviours is developed (MVP)

Author: Added

## Stages of BDD

* Discovery - ensuring the team have a full understanding of a user requirement, using rules and examples to demonstrate this
* Formulation - transforming the examples into documentation that describes how the product behaves (i.e Gherkin)
* Automation - using the documentation to drive automated tests

## Suggested BDD Approach

* Discovery - Customer Representative brings requirements to the team as user stories (PBIs)
* Discovery - The Example Mapping technique is applied to these stories by either the whole team or a “Three Amigos” subset

> * Discovery - Create rules and examples.. will drive out questions about the story, with the output a list of examples as acceptance criteria

Reviewer: This sentence doesn't read well, maybe this would be better?
```suggestion
* Discovery - By creating rules and examples, you will drive out questions about the story, with the output a list of examples as acceptance criteria
```

Author: Re-worded

These sessions can be short daily sessions through to longer weekly events. Theses need to take place to complete development effectively.

Different colour post-its can be used to represent rules, examples and questions.

### Approach

1. A story is created by the customer representative and presented to the group.
1. The group discuss the story and determine rules that clarify the purpose of the story.
1. Examples are created to demonstrate understanding of the rules.
1. When creating examples it may become clear that further rules are needed.
1. When there are lots of rules this may indicate the story is too big and should be split.
1. Any questions that cannot be answered in the session should be recorded and answered subsequently.
1. If there are lots of questions this may indicate the story is not understood and needs to go back to the customer.

# Resources

## Books

* Discovery: Explore behaviour using examples - Gaspar Nagy
  [Amazon link](https://www.amazon.co.uk/Discovery-Explore-behaviour-using-examples/dp/1983591254/ref=sr_1_2?keywords=bdd+discovery&qid=1572530563&s=books&sr=1-2)

> * BDD in Action: Behavior-driven development for the whole software lifecycle - John Ferguson Smart

Reviewer: Incorrect spelling
```suggestion
* BDD in Action: Behaviour-driven development for the whole software lifecycle - John Ferguson Smart
```

Author: Fixed
